### ISSUE TITLE
We Need Acceleration for Gradient Accumulation in DDP mode

### ISSUE THREAD

**Issue by MagicFrogSJTU** [pytorch/pytorch#43201]
## 🚀 Feature
We need an option to temporarily close DDP `all_reduce`.

## Motivation
The training speed can be accelerated when combining DDP and gradient accumulation.   
When applying gradient accumulation, the `optimizer.step()` is called every K steps intead of every step. And as we know every training step (with `loss.backward()`) DDP would activate an `all_reduce`. Therefore, there are K-1 steps of `all_reduce` wasted! Considering the time cost of `all_reduce` especially for multi-node  training, the training speed can be accelerated a lot if we supply an option to call DDP `all_reduce` every K step.

## Pitch

```
model = DDP(model)
for data, label in dataloder:
  optimizer.zero_grad()
  # which I want.
  model.stop_gradient_reduce()
  for i in range(K):
    prediction = model(data)    
    if i == K-1:
        model.start_gradient_reduce()
    loss_fn(prediction, label).backward()
  optimizer.step()
```

## Possible Implementation
I notice `reducer` has a method `prepare_for_backward()`.
```
self.reducer.prepare_for_backward(all_parameters)
```
It seems like if all parameter are set ready here, we can safely jump the `all_reduce`. Or did we need to dig deeper?

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski @jlin27

**Comment by mrshenli** [comment]
Hey @MagicFrogSJTU, is the [`DistributedDataParallel.no_sync()`](https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync) context manager what you are looking for?

**Comment by MagicFrogSJTU** [comment]
> Hey @MagicFrogSJTU, is the [`DistributedDataParallel.no_sync()`](https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync) context manager what you are looking for?

Yes! It's what I want.   
Sorry I didn't notice it before.  
Thanks for your time!  

By the way, is it possible that we improve the DDP tutorial so that the others could be informed when they come across such problems?

**Comment by mrshenli** [comment]
> By the way, is it possible that we improve the DDP tutorial so that the others could be informed when they come across such problems?

Good point! We should. I self-assigned to this issue and will add it to the next release.

**Comment by MagicFrogSJTU** [comment]
> > By the way, is it possible that we improve the DDP tutorial so that the others could be informed when they come across such problems?
> 
> Good point! We should. I self-assigned to this issue and will add it to the next release.

Great!

**Comment by Howell-Yang** [comment]
> Hey @MagicFrogSJTU, is the [`DistributedDataParallel.no_sync()`](https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync) context manager what you are looking for?

I applied model.no_sync() to my code, but find the allreduce wasn't skipped. From the doc [here](https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync), the allreduce process is executed  in the later forward func;
 
"A context manager to disable gradient synchronizations across DDP processes. Within this context, gradients will be accumulated on module variables, which will later be synchronized in the first forward-backward pass exiting the context."

So, we need another solution.
